src/relay/interpreter.cc

```diff
@@ -0,0 +1,459 @@
+
+    setter(arg_len - 1, out_tensor->data);
+    func.CallPacked(TVMArgs(values.data(), codes.data(), arg_len), &ret);
+    return out_tensor;
+  }
+
+  Value Invoke(const Closure& closure, const tvm::Array<Value>& args) {
+    // Get a reference to the function inside the closure.
+    auto func = closure->func;
+    auto compiled = operator_map_.find(func);
+    tvm::Array<Function> funcs;
+    for (auto op : operator_map_) {
+      funcs.push_back(op.first);
+
+    }
+
+    // This case we know we have precompiled the operator.
+    if (compiled != operator_map_.end()) {
+      auto func_ty = func->func_type_annotation();
+      return InvokeCompiledOp(compiled->second, args, func_ty->ret_type);
+    }
+
+  // Allocate a frame with the parameters and free variables.
+    tvm::Map<Var, Value> locals;
+
+    CHECK(func->params.size() == args.size());
+
+    for (size_t i = 0; i < func->params.size(); i++) {
+      CHECK_EQ(locals.count(func->params[i]), 0);
+      locals.Set(func->params[i], args[i]);
+    }
+
+    // Add the var to value mappings from the Closure's environment.
+    for (auto it = closure->env.begin(); it != closure->env.end(); ++it) {
+      CHECK_EQ(locals.count((*it).first), 0);
+      locals.Set((*it).first, (*it).second);
+    }
+
+    // RELAY_DEBUG("invoke", locals, func->body);
+
+    return with_frame<Value>(Frame(locals), [&]() { return Eval(func->body); });
+  }
+
+  Value VisitExpr_(const CallNode* call) override {
+    tvm::Array<Value> args;
+    for (auto arg : call->args) {
+      args.push_back(Eval(arg));
+    }
+
+    // We should not find operators after running fusion,
+    // and operator lowering.
+    //
+    // We have some functions cotaining chunks of operators
+    // which will be loaded into operator map.
+    if (auto op_node = call->op.as<OpNode>()) {
+      LOG(FATAL) << "found " << op_node->name << "; operators should be removed by future passes; try fusing and lowering";
+    }
+
+    // Now we just evaluate and expect to find a closure.
+    Value fn_val = Eval(call->op);
+    if (const ClosureNode* closure_node = fn_val.as<ClosureNode>()) {
+      auto closure = GetRef<Closure>(closure_node);
+      return this->Invoke(closure, args);
+    } else {
+        throw EvalError(
+            "internal error: type error, expected function value in the call position");
+    }
+  }
+
+  Value VisitExpr_(const LetNode* op) override {
+    auto value = Eval(op->value);
+    this->extend(op->var, value);
+    return Eval(op->body);
+  }
+
+  Value VisitExpr_(const TupleGetItemNode* op) override {
+    Value val = Eval(op->tuple);
+    if (auto product_node = val.as<TupleValueNode>()) {
+      return product_node->fields[op->index];
+    } else {
+      throw EvalError("not a product");
+    }
+  }
+
+  Value VisitExpr_(const IfNode* op) override {
+    Value v = Eval(op->cond);
+    if (const TensorValueNode* bv = v.as<TensorValueNode>()) {
+      // RELAY_DEBUG(v);
+      // TODO(@jroesch): Ask TQ
```

Comment:
look at deadcode.cc, there is similar thing
I think we should share them, so dc and eval has same behavior

Reply:
Could you open small PR refactoring them? can open some concurrency up then I could patch here?

Reply:
sure, after you merge?

Reply:
or you want them now?

Reply:
Can you just open a PR after we merge.